hugr-core/src/hugr/views.rs

```diff
@@ -557,57 +552,106 @@ impl<T: AsRef<Hugr>> HugrView for T {
             .map(|(p1, p2)| {
-                [p1, p2].map(|link| hugr.graph.port_offset(link.port()).unwrap().into())
+                [p1, p2].map(|link| self.graph.port_offset(link.port()).unwrap().into())
             })
     }
 
     #[inline]
     fn num_ports(&self, node: Node, dir: Direction) -> usize {
-        self.as_ref().graph.num_ports(node.pg_index(), dir)
+        self.graph.num_ports(node.pg_index(), dir)
     }
 
     #[inline]
     fn children(&self, node: Node) -> impl DoubleEndedIterator<Item = Node> + Clone {
-        self.as_ref().hierarchy.children(node.pg_index()).map_into()
+        self.hierarchy.children(node.pg_index()).map_into()
     }
 
     #[inline]
     fn neighbours(&self, node: Node, dir: Direction) -> impl Iterator<Item = Node> + Clone {
-        self.as_ref()
-            .graph
-            .neighbours(node.pg_index(), dir)
-            .map_into()
+        self.graph.neighbours(node.pg_index(), dir).map_into()
     }
 
     #[inline]
     fn all_neighbours(&self, node: Node) -> impl Iterator<Item = Node> + Clone {
-        self.as_ref()
-            .graph
-            .all_neighbours(node.pg_index())
-            .map_into()
+        self.graph.all_neighbours(node.pg_index()).map_into()
+    }
+}
+
+macro_rules! hugr_view_methods {
+    () => {
+        delegate! {
+            to (**self) {
+
+                #[inline]
+                fn contains_node(&self, node: Node) -> bool;
+
+                #[inline]
+                fn node_count(&self) -> usize;
+
+                #[inline]
+                fn edge_count(&self) -> usize;
+
+                #[inline]
+                fn nodes(&self) -> impl Iterator<Item = Node> + Clone;
+
+                #[inline]
+                fn node_ports(&self, node: Node, dir: Direction) -> impl Iterator<Item = Port> + Clone;
+
+                #[inline]
+                fn all_node_ports(&self, node: Node) -> impl Iterator<Item = Port> + Clone;
+
+                #[inline]
+                fn linked_ports(
+                    &self,
+                    node: Node,
+                    port: impl Into<Port>,
+                ) -> impl Iterator<Item = (Node, Port)> + Clone;
+
+                #[inline]
+                fn node_connections(&self, node: Node, other: Node) -> impl Iterator<Item = [Port; 2]> + Clone;
+
+                #[inline]
+                fn num_ports(&self, node: Node, dir: Direction) -> usize;
+
+                #[inline]
+                fn children(&self, node: Node) -> impl DoubleEndedIterator<Item = Node> + Clone;
+
+                #[inline]
+                fn neighbours(&self, node: Node, dir: Direction) -> impl Iterator<Item = Node> + Clone;
+
+                #[inline]
+                fn all_neighbours(&self, node: Node) -> impl Iterator<Item = Node> + Clone;
+            }
+        }
     }
 }
 
+impl<T: HugrView> HugrView for &T {
+    hugr_view_methods! {}
+}
+
+impl<T: HugrView> HugrView for &mut T {
+    hugr_view_methods! {}
+}
```

Comment:
- There's no need to add `inline`s. `derive_more` takes care of that already.
- We normally single-line the methods when using `derive!`
- With the above, I don't think the macro simplifies the code much.
  Writing it explicitly avoids the extra layer of indirection.

```suggestion
impl<T: HugrView> HugrView for &T {
    delegate! {
        to (**self) {
            fn contains_node(&self, node: Node) -> bool;
            fn node_count(&self) -> usize;
            fn edge_count(&self) -> usize;
            fn nodes(&self) -> impl Iterator<Item = Node> + Clone;
            fn node_ports(&self, node: Node, dir: Direction) -> impl Iterator<Item = Port> + Clone;
            fn all_node_ports(&self, node: Node) -> impl Iterator<Item = Port> + Clone;
            fn linked_ports(&self, node: Node, port: impl Into<Port>) -> impl Iterator<Item = (Node, Port)> + Clone;
            fn node_connections(&self, node: Node, other: Node) -> impl Iterator<Item = [Port; 2]> + Clone;
            fn num_ports(&self, node: Node, dir: Direction) -> usize;
            fn children(&self, node: Node) -> impl DoubleEndedIterator<Item = Node> + Clone;
            fn neighbours(&self, node: Node, dir: Direction) -> impl Iterator<Item = Node> + Clone;
            fn all_neighbours(&self, node: Node) -> impl Iterator<Item = Node> + Clone;
        }
    }
}

impl<T: HugrView> HugrView for &mut T {
    delegate! {
        to (**self) {
            fn contains_node(&self, node: Node) -> bool;
            fn node_count(&self) -> usize;
            fn edge_count(&self) -> usize;
            fn nodes(&self) -> impl Iterator<Item = Node> + Clone;
            fn node_ports(&self, node: Node, dir: Direction) -> impl Iterator<Item = Port> + Clone;
            fn all_node_ports(&self, node: Node) -> impl Iterator<Item = Port> + Clone;
            fn linked_ports(&self, node: Node, port: impl Into<Port>) -> impl Iterator<Item = (Node, Port)> + Clone;
            fn node_connections(&self, node: Node, other: Node) -> impl Iterator<Item = [Port; 2]> + Clone;
            fn num_ports(&self, node: Node, dir: Direction) -> usize;
            fn children(&self, node: Node) -> impl DoubleEndedIterator<Item = Node> + Clone;
            fn neighbours(&self, node: Node, dir: Direction) -> impl Iterator<Item = Node> + Clone;
            fn all_neighbours(&self, node: Node) -> impl Iterator<Item = Node> + Clone;
        }
    }
}
```

Reply:
> * There's no need to add `inline`s. `derive_more` takes care of that already.

Sounds good, but how does `derive_more` get involved here? I took the `[inline]`s from the old `impl<T: AsRef<Hugr>> HugrView for T {`...

I've now managed to common up all three of the delegations with the same macro...

Reply:
Sorry, _s/derive_more/delegate/g_ -.-'

Reply:
Ah....right. (That makes sense!)

Ok, so if `delegate` automatically adds `[inline]`....then (a) that shortens the macro :), and (b) yes it does reduce the utility of the macro a bit. OTOH there are now three instances of it, so...we can drop it....but do you really want to? :)

Note the macro is local to `hugr::views`, which I think works quite well, but we'd have to put `Rc` / `Cow` / `Arc` versions thereof in `views::refs` or something - not a bad place I guess. If we do those as well it'd be 5-6 instances of the macro.... (but they are easy to add, even in this PR)

Reply:
Done!